### ISSUE TITLE
Adding CRDs to a cluster causes deadlock

### ISSUE THREAD

**Issue by jdroot** [argoproj/argo-cd#18902]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

This bug exists entirely inside of [pkg/cache/cluster.go](https://github.com/argoproj/gitops-engine/blob/master/pkg/cache/cluster.go). The deadlock can be seen in my goroutines dump from pprof: [goroutines.zip](https://github.com/user-attachments/files/16074251/goroutines.zip)

It is the classic case of two code paths attempting to grab two locks in different orders. Resulting in each code path holding one side, unable to grab the other, and unable to progress to releasing what they hold.

**First code path**
`loadInitialState` gets called over 500 times in our Argo instances, with `lock` equal to `true`. This leads in to a call to `listResources` which starts by grabbing a limited Semaphore, only `defaultListSemaphoreWeight` Go routines can do this at any one time. `listResources` then invokes the callback supplied on line 571. This callback, when `lock` is `true`, attempts to grab `c.lock` (line 586).

**Second code path**
When a CRD is added to the cluster, it gets processed on line 662, inside `watchEvents`. On line 688 we see an `if` statement for the event type, the `else` will fire when a CRD is `ADDED` (or `MODIFIED`, but I have not verified that case). When the `else` runs, it ends up grabbing `c.lock` with `runSynced` and then calling `startMissingWatches`. This eventually calls `loadInitialState`, which calls `listResources`, which attempts to grab the semaphore.

When the second code path runs while there are more than `defaultListSemaphoreWeight` go routines calling `listResources` (which I believe is always), it grabs `c.lock` and waits on the semaphore in `listResources`. The other `listResources` code path is holding the semaphore and waiting on `c.lock`.

In the goroutine dump, we would expect, and I can see, that there are exactly `defaultListSemaphoreWeight + 1` callers to `runSync`. We can see that `defaultListSemaphoreWeight` of them are waiting on `c.lock` and one is waiting on `listSemaphore.Acquire` while holding `c.lock`.

**To Reproduce**
This issue happens when the number of Resource types in a cluster exceeds `defaultListSemaphoreWeight`. It is unclear if only 51 are needed or just that additional ones increase the likelihood of the deadlock triggering. In our case, we have over 500 CRDs and the issue triggers every time. If it is a chance to trigger, lowering `defaultListSemaphoreWeight` should also increase the odds.

1. Add a CRD to the cluster
2. Observe that Argo becomes "wedged" on that cluster and unable to make progress
3. Run `kubectl port-forward -n <argos-namespace> <argo controller pod> 8082:8082
4. Browse to [http://localhost:8082/debug/pprof/goroutine?debug=2](http://localhost:8082/debug/pprof/goroutine?debug=2) to see all Go routines
5. Observe that there are `defaultListSemaphoreWeight + 1` go routines calling `runSynced`, one of which is waiting on the semaphores `Acquire` function.

**Expected behavior**

I would expect ArgoCD to properly handle the addition (or modification) of new CRDs

**Version**
ArgoCD v2.10.11
The bug is in `gitops-engine`, our argo was compiled from SHA `fbecbb86e41254a75a59943b5eb43ed55d21cdb9`

**Logs**
`startMissingWatches` go routine:
```
goroutine 4014546 [select, 5535 minutes]:
golang.org/x/sync/semaphore.(*Weighted).Acquire(0xc0057b2000, {0x581e7e0, 0xc01bd35720}, 0x1)
	/go/pkg/mod/golang.org/x/sync@v0.3.0/semaphore/semaphore.go:60 +0x314
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).listResources(0xc0057aa300, {0x581e7e0?, 0xc01bd35720?}, {0x58339d0?, 0xc01bd35770}, 0xc023718be8)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:531 +0x86
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).loadInitialState(_, {_, _}, {{{0xc045082708, 0x15}, {0xc031eea6c0, 0x18}}, {{0xc0271f5380, 0x19}, {0xc031eea6d8, ...}, ...}, ...}, ...)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:571 +0x93
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).startMissingWatches.func1({0x58339d0?, 0xc01bd35770}, {0x0, 0x0})
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:494 +0xd0
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).processApi(_, {_, _}, {{{0xc045082708, 0x15}, {0xc031eea6c0, 0x18}}, {{0xc0271f5380, 0x19}, {0xc031eea6d8, ...}, ...}, ...}, ...)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:736 +0xca
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).startMissingWatches(0xc0057aa300)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:493 +0x370
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).watchEvents.func1.4()
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:701 +0x17
github.com/argoproj/gitops-engine/pkg/cache.runSynced({0x57f5ac0, 0xc0057aa3a0}, 0xc01f827560)
```

Typical `loadInitialState` go routine:
```
goroutine 4014691 [sync.Mutex.Lock, 5535 minutes]:
sync.runtime_SemacquireMutex(0xc001088ed0?, 0xab?, 0x57f56b0?)
	/usr/local/go/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0057aa3a0)
	/usr/local/go/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:90
sync.(*RWMutex).Lock(0x0?)
	/usr/local/go/src/sync/rwmutex.go:147 +0x31
github.com/argoproj/gitops-engine/pkg/cache.runSynced({0x57f5ac0, 0xc0057aa3a0}, 0xc001089010)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:524 +0x3c
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).loadInitialState.func1(0xc0057b2000?)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:586 +0x307
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).listResources(0xc0057aa300, {0x581e7e0?, 0xc0294a84b0?}, {0x58339d0?, 0xc0294a8500}, 0xc0010891a0)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:567 +0x202
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).loadInitialState(_, {_, _}, {{{0xc028dc00e0, 0x1c}, {0xc0305d3210, 0x9}}, {{0xc0305d3200, 0xb}, {0xc0305d3230, ...}, ...}, ...}, ...)
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:571 +0x93
github.com/argoproj/gitops-engine/pkg/cache.(*clusterCache).watchEvents.func1()
	/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240416142647-fbecbb86e412/pkg/cache/cluster.go:607 +0x156
```

I believe this bug report details the cause of bug reports like [16950](https://github.com/argoproj/argo-cd/issues/16950)

**Comment by christianh814** [comment]
@jdroot Just for completeness, does this also occur when you use [server side apply](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#server-side-apply) within the Application settings?

**Comment by jdroot** [comment]
@christianh814 On the Application creating the CRD? Or where do you mean. Based on my reading of the code I don't believe that setting is involved at all. The deadlock is entirely isolated to the `gitops-engine`'s tracking of resources

**Comment by christianh814** [comment]
@jdroot My apologies. I was under the assumption that when you said "Add a CRD to the cluster" you mean an Argo CD Application applying/installing a CRD.

What do you mean by "Add a CRD to the cluster"? Do you mean adding the Argo CD CRDs or something else?

**Comment by jdroot** [comment]
@christianh814 I specifically mean a new custom resource type getting installed in the cluster. Something that would show up when running `kubectl get crds`. In our case, Argo **was** used to install the CRD that triggered the deadlock, but the new CRDs broke all ArgoCD instances monitoring the cluster the CRD was installed in to.

**Comment by jdroot** [comment]
I spent the day trying to reproduce this in one of our clusters by applying and removing lots of CRDs and was unable to. Based on reading the code (opposing lock orders from two different go-routines) and the go-routine dump proving that is what happened I am absolutely confident the bug exists, but I think the reproduction steps might be more complicated.

For a little bit on our setup, maybe letting someone else realize the direct trigger.. We have many clusters (20+). Each cluster has an ArgoCD instance running in it configured to only work with in-cluster. We also have many (10+) ArgoCD instances running in a management cluster with the ability to deploy to all clusters. Every cluster has 500+ CRDs, 50,000+ objects, and thousands of ArgoCD Applications targeting them, spread across all of our ArgoCD instances. We applied about 60 new CRDs to the clusters last week, at the same time in every cluster, and nearly all of the ArgoCD controllers became deadlocked in the way described in my initial report. This leads me to the conclusion that this is not a very rare condition to trigger, but it is also not deterministic. I imagine there is some additional internal or external state I am not understanding that actually causes this to trigger.


The submitted PR adds a bool to allow callers to skip the semaphore. The semaphore was used for rate limiting, so this does open up the potential for more callers to be running at once, but in practice that should be limited to just 1 additional caller. That seems like a fine price to pay to avoid deadlock
